File: checksum/get_key/solve.c
#include <stdio.h>
#include <stdint.h>

#define CHECKSUM_SIZE 0x40
#define KEY "FlareOn2024"
#define HIGHQ(x) ((int64_t)(((__uint128_t)(x) * (__uint128_t)0x5d1745d1745d1746) >> 64))

int main() {
    uint8_t checksum[CHECKSUM_SIZE] = {0};
    const int64_t constant = 0x5d1745d1745d1746;
    const uint8_t want[CHECKSUM_SIZE] = {
        0x71, 0x0A, 0x05, 0x45, 0x01, 0x2B, 0x5F, 0x56,
        0x00, 0x57, 0x0D, 0x73, 0x55, 0x07, 0x45, 0x51,
        0x2C, 0x5F, 0x01, 0x03, 0x51, 0x05, 0x75, 0x0D,
        0x03, 0x10, 0x52, 0x7B, 0x5E, 0x50, 0x09, 0x54,
        0x55, 0x27, 0x5A, 0x50, 0x13, 0x07, 0x7F, 0x58,
        0x50, 0x54, 0x02, 0x51, 0x25, 0x08, 0x50, 0x45,
        0x52, 0x79, 0x5A, 0x07, 0x55, 0x0B, 0x07, 0x24,
        0x5D, 0x04, 0x41, 0x5D, 0x7D, 0x5B, 0x56, 0x54
    };

    for (int64_t idx = 0; idx < CHECKSUM_SIZE; idx++) {
        int64_t mul = HIGHQ(idx);
        int64_t key_idx = idx - ((mul >> 2) * 0xB);
        fprintf(stderr, "idx=%lld, mul=%lld, key_idx=%lld\n", idx, mul, key_idx);

        // Check if key_idx is within bounds
        if (key_idx < 0 || key_idx >= 11) {
            fprintf(stderr, "Index out of bounds: key_idx=%lld, idx=%lld\n", key_idx, idx);
            return 1;  // Exit with an error
        }

        checksum[idx] = want[idx] ^ KEY[key_idx];
        printf("%c", checksum[idx]);
    }

    // Print the checksum as a string
    for (int i = 0; i < CHECKSUM_SIZE; i++) {
        printf("%c", checksum[i]);
    }
    printf("\n");

    return 0;
}

File: checksum/get_key/solve.py
QWORD_SIZE = 2**64

# unknown
checksum_size = 0x20
key = b"FlareOn2024"
checksum = [0 for _ in range(checksum_size)]

const = 0x5d1745d1745d1746
want = b"\x71\n\x05\x45\x01\x2b\x5f\x56\0\x57\r\x73\x55\x07\x45\x51\x2c\x5f\x01\x03\x51\x05\x75\r\x03\x10\x52\x7b\x5e\x50\t\x54\x55\'\x5a\x50\x13\x07\x7f\x58\x50\x54\x02\x51\x25\b\x50\x45\x52\x79\x5a\x07\x55\x0b\x07\x24\x5d\x04\x41\x5d\x7d\x5b\x56\x54"

idx = 0
while idx <= checksum_size:
    mul = (const * idx) % QWORD_SIZE
    key_idx = (idx - ((mul >> 2) * 0xb)) % QWORD_SIZE
    print(key_idx)
    checksum[idx] = want[idx] ^ key[key_idx]
    print(chr(checksum[idx]))
    idx+=1

print("".join([chr(x) for x in checksum]))
